cmake/toolchain_llvm.cmake

```diff
@@ -11,13 +11,42 @@ set(CMAKE_SYSTEM_NAME Generic)
 
 set(LLVM_TAG llvm)
 
-set(CMAKE_C_COMPILER ${TOOLCHAIN_DIR}/clang)
-set(CMAKE_CXX_COMPILER ${TOOLCHAIN_DIR}/clang++)
-set(CMAKE_ASM_COMPILER ${TOOLCHAIN_DIR}/clang)
+set(CMAKE_C_COMPILER ${TOOLCHAIN_DIR}/bin/clang)
+set(CMAKE_CXX_COMPILER ${TOOLCHAIN_DIR}/bin/clang++)
+set(CMAKE_ASM_COMPILER ${TOOLCHAIN_DIR}/bin/clang)
 
-set(CMAKE_OBJCOPY ${TOOLCHAIN_DIR}/${LLVM_TAG}-objcopy)
-set(CMAKE_OBJDUMP ${TOOLCHAIN_DIR}/${LLVM_TAG}-objdump)
-set(CMAKE_AR ${TOOLCHAIN_DIR}/${LLVM_TAG}-ar)
+set(CMAKE_OBJCOPY ${TOOLCHAIN_DIR}/bin/${LLVM_TAG}-objcopy)
+set(CMAKE_OBJDUMP ${TOOLCHAIN_DIR}/bin/${LLVM_TAG}-objdump)
+set(CMAKE_AR ${TOOLCHAIN_DIR}/bin/${LLVM_TAG}-ar)
+
+# Disable ABI detection
+set(CMAKE_C_ABI_COMPILED "False")
 
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} --target=riscv32-unknown-elf")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --target=riscv32-unknown-elf")
\ No newline at end of file
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --target=riscv32-unknown-elf")
+
+# Use LLVM LLD linker
+set(CMAKE_EXE_LINKER_FLAGS "-fuse-ld=lld")
+
+# Check LLVM version
+execute_process(
+    COMMAND ${CMAKE_C_COMPILER} --version
+    OUTPUT_VARIABLE LLVM_VERSION
+)
+string(REGEX MATCH "[0-9]+\\.[0-9]+\\.[0-9]+" LLVM_VERSION ${LLVM_VERSION})
+
+# Get Major version
+string(REGEX MATCH "^[0-9]+" LLVM_VERSION_MAJOR ${LLVM_VERSION})
+string(REGEX MATCH "[0-9]+$" LLVM_VERSION_MINOR ${LLVM_VERSION})
+string(REGEX MATCH "[0-9]+$" LLVM_VERSION_PATCH ${LLVM_VERSION})
+
+if (LLVM_VERSION_MAJOR LESS 15)
+    message(STATUS "Disable linker relaxation for LLVM < 15")
+    # WIESEP: Disable linker relaxation for LLVM 12
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mno-relax")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mno-relax")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--no-relax")
+
+    # Also for ASM files
+    set(CMAKE_ASM_FLAGS "${CMAKE_ASM_FLAGS} -mno-relax")
```

Comment:
Same point here - I would use `add_compile_options` and `add_link_options` instead - setting variables directly is imo more confusing.

Reply:
Good point, I changed this.